### PR TITLE
Kafka clarifications 

### PR DESCRIPTION
**Pull request by clemensv**
Related to #571 

Signed-off-by: clemensv <clemensv@microsoft.com>

### CODE REVIEW

File: kafka-protocol-binding.md [hunk 1]
```diff
@@ -50,6 +50,18 @@ This specification does not prescribe rules constraining transfer or settlement
 of event messages with Kafka; it solely defines how CloudEvents are expressed
 in the Kafka protocol as [Kafka messages][kafka-message-format].
 
+The Kafka documentation uses "message" and "record" somewhat interchangeably
+and therefore the terms should be considered being synonyms for this spec as
```
Comment (duglin) [inline kafka-protocol-binding.md:54]: s/spec/specification/

File: kafka-protocol-binding.md [hunk 2]
```diff
@@ -130,14 +142,19 @@ this specification then it is not a valid CloudEvent.
 If the `content-type` header is not present then the receiver uses
 _structured_ mode with the JSON event format.
 
-### 3.1. Key Attribute
+### 3.1. Key mapping
```
Comment (duglin) [inline kafka-protocol-binding.md:145]: s/mapping/Mapping

File: kafka-protocol-binding.md (hunk 1 above)
Comment (duglin) [reply kafka-protocol-binding.md:54]: s/should/are to/

File: kafka-protocol-binding.md [hunk 3]
```diff
@@ -130,14 +142,19 @@ this specification then it is not a valid CloudEvent.
...
 
-The 'key' attribute is populated by a partitionKeyExtractor function. The
-partitionKeyExtractor is a protocol specific function that contains bespoke
-logic to extract and populate the value. The key attribute MUST be encoded
-as UTF-8 string both in the Kafka message and in the CloudEvents extension.
-A default implementation of the extractor will use the
-[Partitioning](extensions/partitioning.md) extension value.
+Every implementation SHOULD provide a default "Key Mapper" implementation that
+maps the [Partitioning](extensions/partitioning.md) `partitionkey` attribute
+value to the 'key' of the Kafka message as-is, if present.
```
Comment (slinkydeveloper) [inline kafka-protocol-binding.md:157]: Other than writing the value of `partitionKey` as `key` of kafka message, should the implementation write the `partitionKey` as extension too? Or it applies the same rule of `content-type` (aka i write content-type but not `ce_datacontenttype`) ?

File: kafka-protocol-binding.md [hunk 4]
```diff
@@ -130,14 +142,19 @@ this specification then it is not a valid CloudEvent.
...
+specific.
 
-The 'key' attribute is populated by a partitionKeyExtractor function. The
-partitionKeyExtractor is a protocol specific function that contains bespoke
-logic to extract and populate the value. The key attribute MUST be encoded
-as UTF-8 string both in the Kafka message and in the CloudEvents extension.
-A default implementation of the extractor will use the
-[Partitioning](extensions/partitioning.md) extension value.
+Every implementation SHOULD provide a default "Key Mapper" implementation that
+maps the [Partitioning](extensions/partitioning.md) `partitionkey` attribute
```
Comment (slinkydeveloper) [inline kafka-protocol-binding.md:156]: `partitionKey` doesn't follow the naming convention of attributes in CloudEvents spec: https://github.com/cloudevents/spec/blob/v1.0/spec.md#attribute-naming-convention

**Comment by slinkydeveloper** [comment]
Implementation in sdk-go: https://github.com/cloudevents/sdk-go/pull/437

File: kafka-protocol-binding.md (hunk 3 above)
Comment (clemensv) [reply kafka-protocol-binding.md:157]: If the value is part of the event, it must also be written into the event. A different key mapper function could take OOB information and use that for the key. I'll add a clarification.

Comment (slinkydeveloper) [reply kafka-protocol-binding.md:157]: Okay so the event, in that case, is left unchanged.

File: kafka-protocol-binding.md (hunk 4 above)
Comment (clemensv) [reply kafka-protocol-binding.md:156]: Yes. It must be. We need to fix that.

File: kafka-protocol-binding.md [hunk 5]
```diff
@@ -130,14 +142,25 @@ this specification then it is not a valid CloudEvent.
...
 
-The 'key' attribute is populated by a partitionKeyExtractor function. The
-partitionKeyExtractor is a protocol specific function that contains bespoke
-logic to extract and populate the value. The key attribute MUST be encoded
-as UTF-8 string both in the Kafka message and in the CloudEvents extension.
-A default implementation of the extractor will use the
-[Partitioning](extensions/partitioning.md) extension value.
+A mapping function MUST NOT modify the CloudEvent. This means that the
+aforementioned `partitionkey` attribute MUST still be included with the
+transmitted event. It also means that a mapping function that uses key
```
Comment (slinkydeveloper) [inline kafka-protocol-binding.md:161]: > transmitted event.

transmitted event, if present.

Comment (clemensv) [reply kafka-protocol-binding.md:161]: That condition is still in the paragraph above, no?

Comment (slinkydeveloper) [reply kafka-protocol-binding.md:161]: It doesn't hurt to repeat IMO, @duglin?

**Comment by duglin** [comment]
@clemensv I think you pulled in one of subscription API docs by mistake

File: kafka-protocol-binding.md [hunk 6]
```diff
@@ -130,14 +142,25 @@ this specification then it is not a valid CloudEvent.
...
+value to the 'key' of the Kafka message as-is, if present. 
 
-The 'key' attribute is populated by a partitionKeyExtractor function. The
-partitionKeyExtractor is a protocol specific function that contains bespoke
-logic to extract and populate the value. The key attribute MUST be encoded
-as UTF-8 string both in the Kafka message and in the CloudEvents extension.
-A default implementation of the extractor will use the
-[Partitioning](extensions/partitioning.md) extension value.
+A mapping function MUST NOT modify the CloudEvent. This means that the
+aforementioned `partitionkey` attribute MUST still be included with the
```
Comment (duglin) [inline kafka-protocol-binding.md:160]: Does this make the partitionKey attribute a required attribute now?

**Comment by duglin** [comment]
Approved on 4/16 call

**Pull request merged**